src/coreclr/src/ilasm/portable_pdb.cpp

```diff
@@ -0,0 +1,396 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "portable_pdb.h"
+#include <time.h>
+#include "assembler.h"
+
+//*****************************************************************************
+// Document
+//*****************************************************************************
+Document::Document()
+{
+    m_name = NULL;
+    m_token = mdDocumentNil;
+}
+
+Document::~Document()
+{
+    if (m_name)
+    {
+        delete[] m_name;
+        m_name = NULL;
+    }
+    m_token = mdDocumentNil;
+};
+
+char* Document::GetName()
+{
+    return m_name;
+}
+
+void Document::SetName(char* name)
+{
+    m_name = new char[strlen(name) + 1];
+    strcpy_s(m_name, strlen(name) + 1, name);
+}
+
+mdDocument Document::GetToken()
+{
+    return m_token;
+}
+
+void Document::SetToken(mdDocument token)
+{
+    m_token = token;
+}
+
+
+//*****************************************************************************
+// PortablePdbWritter
+//*****************************************************************************
+PortablePdbWritter::PortablePdbWritter()
+{
+    m_pdbStream.id.pdbGuid = { 0 };
+    m_pdbStream.id.pdbTimeStamp = 0;
+    m_pdbStream.entryPoint = mdMethodDefNil;
+    m_pdbStream.referencedTypeSystemTables = 0UL;
+    m_pdbStream.typeSystemTableRows = new ULONG[TBL_COUNT];
+    m_pdbStream.typeSystemTableRowsSize = 0;
+    m_currentDocument = NULL;
+    m_pdbEmitter = NULL;
+}
+
+PortablePdbWritter::~PortablePdbWritter()
+{
+    if (m_pdbEmitter != NULL)
+    {
+        m_pdbEmitter->Release();
+        m_pdbEmitter = NULL;
+    }
+    if (m_pdbStream.typeSystemTableRows != NULL)
+    {
+        delete[] m_pdbStream.typeSystemTableRows;
+        m_pdbStream.typeSystemTableRows = NULL;
+    }
+
+    m_documentList.RESET(true);
+}
+
+HRESULT PortablePdbWritter::Init(IMetaDataDispenserEx2* mdDispenser)
+{
+    HRESULT hr = S_OK;
+    if (m_pdbEmitter != NULL)
+    {
+        m_pdbEmitter->Release();
+        m_pdbEmitter = NULL;
+    }
+    m_currentDocument = NULL;
+    m_documentList.RESET(true);
+
+    memset(m_pdbStream.typeSystemTableRows, 0, sizeof(ULONG) * TBL_COUNT);
+    time_t now;
+    time(&now);
+    m_pdbStream.id.pdbTimeStamp = (ULONG)now;
```

Comment:
Do we really want the timestamp in the generated PDB file? That makes the generated file not binary equivalent on equivalent source rebuild. I thought we're trying to get away from timestamps in generated files because of this problem.